lib/demo/interactors/create_comment.rb

```diff
@@ -0,0 +1,23 @@
+require 'lotus/interactor'
+
+module Demo
+  class CreateComment
+    include Lotus::Interactor
+
+    def initialize(params:, repository: CommentRepository)
+      @repository = repository
+      @params = params
+    end
+
+    def call
```

Comment:
The usage, of `#error!` has the effect of set an error message, and to terminate the execution of `#call`. The goal is to not proceed with the other steps, if there is an error that can't be handled (eg. persistence exception).

In the case above, if the comment cannot be persisted, there is no reason to deliver notifications and so on..

On the other hand, `#error` collects the message, but doesn't terminate the flow. We can easily understand that `#error` can be invoked multiple times, while its "bang counterpart" can't.

### error example

```ruby
def call
  validate_params
  persist_comment
end

private
def validate_params
  unless @params.valid?
    @params.errors.each {|err| error(err.message) }
  end
end
```

In the example above, we can invoke `#error` multiple time so we can expose messages to the users (via `result.errors`). Please note that `#persist_comment` will be invoked.

### error! example

```ruby
def call
  verify_preconditions
  persist_comment
end

private
def verify_preconditions
  unless can_reply_to_post?
    error! "You can't reply to the post"
  end
end
```

In the case above `#persist_comment` won't be called.

### error and fail!

If you want to collect multiple errors and then terminate the execution you can use `#error` in combination with `#fail!`

```ruby
def call
  validate_params
  persist_comment
end

private
def validate_params
  unless @params.valid?
    @params.errors.each {|err| error(err.message) }
    fail!
  end
end
```

In the case above, `#persist_comment` won't be invoked.